pcs/lib/corosync/config_facade.py

```diff
@@ -578,66 +609,82 @@ def set_totem_options(self, options):
         # Note: all options in totem section supported by pcs are runtime
         # configurable.
 
-    def get_totem_options(self) -> Dict[str, str]:
+    def get_totem_options(self) -> dict[str, str]:
         """
         Get configurable totem options
         """
         return self._filter_options("totem", constants.TOTEM_OPTIONS)
 
-    def set_quorum_options(self, options):
+    def set_quorum_options(self, options: Mapping[str, str]) -> None:
         """
         Set options in the "quorum" section
 
-        dict options -- quorum options
+        options -- quorum options
         """
         quorum_section_list = self.__ensure_section(self.config, "quorum")
         self.__set_section_options(quorum_section_list, options)
         self.__update_two_node()
         self.__remove_empty_sections(self.config)
         self._need_stopped_cluster = True
 
-    def get_quorum_options(self) -> Dict[str, str]:
+    def get_quorum_options(self) -> dict[str, str]:
         """
         Get configurable options from the "quorum" section
         """
         return self._filter_options("quorum", constants.QUORUM_OPTIONS)
 
-    def is_enabled_auto_tie_breaker(self):
+    def is_enabled_auto_tie_breaker(self) -> bool:
         """
         Returns True if auto tie braker option is enabled, False otherwise.
         """
         return self._get_option_value("quorum", "auto_tie_breaker", "0") == "1"
 
-    def has_quorum_device(self):
+    def _get_raw_quorum_device_model(self) -> Optional[str]:
+        models_found = []
+        for quorum in self.config.get_sections("quorum"):
+            for device in quorum.get_sections("device"):
+                model_list = device.get_attributes("model")
+                if model_list:
+                    models_found.append(model_list[-1][1])
+        return models_found[-1] if models_found else None
+
+    def has_quorum_device(self) -> bool:
         """
         Check if quorum device is present in the config
         """
-        for quorum in self.config.get_sections("quorum"):
-            for device in quorum.get_sections("device"):
-                if device.get_attributes("model"):
-                    return True
-        return False
+        return bool(self._get_raw_quorum_device_model())
 
-    def get_quorum_device_model(self):
+    def get_quorum_device_model(self) -> str:
```

Comment:
The `has_quorum_device` method existed before I started adding type hints.

The proposed change may look like it simplifies things. But I don't think it's worth it. If `get_quorum_device_model` could return `None`, we would have to check for that everywhere it's used. Otherwise `mypy` would rightfully complain if we tried to use `get_quorum_device_model`'s return value anywhere where only `str` is allowed.

Having one method to check if quorum device is defined and another methods to get quorum device configuration looks better to me than having methods to get quorum device configuration or `None` in case no quorum device is defined and checking if they returned `None` or actual data everywhere they're used.

Reply:
So I checked all `has_quorum_device` and `get_quorum_device_model` calls (in `main` branch as I assume there should not be significant differences with this branch). Every time `get_quorum_device_model` is called, `has_quorum_device` is called right before it with some error handling.

Therefore I don't think check for `None` would add any complexity as basically the same check is done everywhere already, only wrapped in a function call. Which means we need to call 2 separate functions for no real benefit.

Actually, I believe having `mypy` checking for not properly handling the case when there is no quorum device is much better then getting an `AssertionError` at runtime.

Reply:
You are right, your proposed version is actually simpler.